src/persist-client/src/write.rs

```diff
@@ -329,56 +437,224 @@ where
+/// Otherwise, a dangling batch will leak and backing blobs will remain in blob storage.
+// TODO(aljoscha): Right now, we require batches to be manually marked as
+// consumed (when they're appended to a batch) or deleted. This is clunky and
+// prone to errors so we should replace it with lazy cleanup once we can use
+// async drop.
+#[derive(Debug)]
+pub struct Batch<K, V, T, D>
+where
+    T: Timestamp + Lattice + Codec64,
+{
+    /// [Description] of updates contained in this batch.
+    pub desc: Description<T>,
+
+    /// Keys to blobs that make up this batch of updates.
+    pub(crate) blob_keys: Vec<String>,
+
+    // These provide a bit more safety against appending a batch with the wrong
+    // type to a shard.
+    _phantom: PhantomData<(K, V, T, D)>,
+}
+
+impl<K, V, T, D> Drop for Batch<K, V, T, D>
+where
+    T: Timestamp + Lattice + Codec64,
+{
+    fn drop(&mut self) {
+        if self.blob_keys.len() > 0 {
+            // warn!(
+            //     "un-consumed Batch, with {} dangling blob keys: {:?}",
+            //     self.blob_keys.len(),
+            //     self.blob_keys
+            // );
+            // WIP: Panic, so that we can tease out places in our existing code
+            // where we need to consume/delete.
+            panic!(
+                "un-consumed Batch, with {} dangling blob keys: {:?}",
+                self.blob_keys.len(),
+                self.blob_keys
+            );
+        }
+    }
+}
+
+impl<K, V, T, D> Batch<K, V, T, D>
+where
+    K: Debug + Codec,
+    V: Debug + Codec,
+    T: Timestamp + Lattice + Codec64,
+    D: Semigroup + Codec64,
+{
+    pub(crate) fn new(desc: Description<T>, blob_keys: Vec<String>) -> Self {
+        Self {
+            desc,
+            blob_keys,
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Marks the blobs that this batch handle points to as consumed, likely
+    /// because they were appended to a shard.
+    ///
+    /// Consumers of a blob need to make this explicit, so that we can log
+    /// warnings in case a batch is not used.
+    pub(crate) fn mark_consumed(&mut self) {
+        self.blob_keys.clear();
+    }
+
+    /// Deletes the blobs that make up this batch from the given blob store and
+    /// marks them as deleted.
+    // WIP: This is hard to call for users of the API because they usually don't
+    // have a Blob.
+    pub async fn delete(
+        &mut self,
+        blob: &Arc<dyn BlobMulti + Send + Sync>,
+    ) -> Result<(), ExternalError> {
+        let deadline = Instant::now() + FOREVER;
+        for key in self.blob_keys.iter() {
+            blob.delete(deadline, key).await?;
+        }
+        self.blob_keys.clear();
+        Ok(())
+    }
+}
+
+/// A builder for [Batches](Batch) that allows adding updates piece by piece and
+/// then finishing it.
+#[derive(Debug)]
+pub struct BatchBuilder<K, V, T, D>
+where
+    T: Timestamp + Lattice + Codec64,
+{
+    size_hint: usize,
+    pub(crate) lower: Antichain<T>,
```

Comment:
I think the `pub(crate)` can be removed from all these fields?

Reply:
I removed all `pub(crate)`/`pub` except `pub desc` on `Batch`. Does that sound alright?

Reply:
`desc` is still pub because the persist user might want to be able to see the bounds? It's a little scary that they could muck around with it (plus the since is sorta meaningless). I'd say move that one non-pub, too, and add upper and lower `pub fn`s to Batch